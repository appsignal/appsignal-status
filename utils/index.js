import dayjs from "dayjs";
import utc from "dayjs/plugin/utc";
dayjs.extend(utc);

export const formatRegion = (region) => {
  return region
    .split(/-| /)
    .map((region) => region.charAt(0).toUpperCase() + region.slice(1))
    .join(" ");
};

export const timeseriesByDay = (timeseries) => {
  return sortedTimeseries(timeseries.slice()).reduce((group, timeserie) => {
    const startOfDay = dayjs(timeserie.timestamp).startOf("day").utc().format();

    let currentGroup = group.filter((item) => item.timestamp === startOfDay)[0];

    if (!currentGroup) {
      currentGroup = {
        timestamp: startOfDay,
        values: {},
      };

      group.push(currentGroup);
    }

    if (timeserie.values === undefined) {
      currentGroup.missingDataPoint = true;
      currentGroup.values = {};
    } else {
      Object.keys(timeserie.values).map((region) => {
        if (!currentGroup.values[region]) currentGroup.values[region] = 0;
        currentGroup.values[region] += timeserie.values[region];
      });
    }

    return group;
  }, []);
};

export const sortedTimeseries = (timeseries) => {
  return timeseries.sort(
    (a, b) => new Date(a.timestamp) - new Date(b.timestamp)
  );
};

export const fillMissingDataPoints = (timeseries, expectedDays) => {
  for (let i = 0; i < expectedDays; i++) {
    const date = dayjs().subtract(i, "day").startOf("day").utc().format();
    const timeSerieForDate = timeseries.filter(
      (item) => item.timestamp === date
    )[0];

    if (timeSerieForDate === undefined) {
      const emptyDataPoint = {
        timestamp: date,
        missingDataPoint: true,
        values: {},
      };
      timeseries.push(emptyDataPoint);
    }
  }
  return sortedTimeseries(timeseries);
};
