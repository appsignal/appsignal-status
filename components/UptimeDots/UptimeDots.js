import PropTypes from "prop-types";

import UptimeDot from "../UptimeDot/UptimeDot";
import { fillMissingDataPoints, timeseriesByDay } from "../../utils";

const UptimeDots = ({ timeseries, regions, threshold }) => {
  const filledTimeseries = fillMissingDataPoints(
    timeseriesByDay(timeseries, regions).slice(-30, timeseries.length),
    30
  );

  return (
    <div className="flex space-x-1">
      {filledTimeseries.slice(-30, filledTimeseries.length).map((timeserie) => (
        <UptimeDot
          key={timeserie.timestamp}
          timeserie={timeserie}
          threshold={threshold}
        />
      ))}
    </div>
  );
};

UptimeDots.propTypes = {
  timeseries: PropTypes.array.isRequired,
  regions: PropTypes.array.isRequired,
  threshold: PropTypes.number,
};

export default UptimeDots;
