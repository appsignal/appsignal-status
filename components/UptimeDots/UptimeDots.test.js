import { render, screen } from "@testing-library/react";

import UptimeDots from "./UptimeDots";
import homepageMock from "../../mocks/monitors/homepage.json";
import { timeseriesByDay } from "../../utils";

const build = (props = {}) => {
  return render(<UptimeDots timeseries={homepageMock.timeseries} {...props} />);
};

describe("UptimeDots", () => {
  test("renders without errors", () => {
    const { container } = build();
    expect(container).toMatchSnapshot();
  });

  test("it renders 30 dots for a given set of timeseries", () => {
    build();

    const dots = screen.getAllByTestId("uptimeDot");
    expect(dots.length).toBe(30);
  });

  test("even with missing days we get 30 uptimeDots", () => {
    const timeseries = homepageMock.timeseries.splice(0, 10);
    build({ timeseries });

    const dots = screen.getAllByTestId("uptimeDot");
    expect(dots.length).toBe(30);
  });

  test("can deal with a timeserie without any values", () => {
    const timeseries = timeseriesByDay(homepageMock.timeseries);
    timeseries[0].values = undefined;

    expect(() => build({ timeseries })).not.toThrowError();
  });
});
