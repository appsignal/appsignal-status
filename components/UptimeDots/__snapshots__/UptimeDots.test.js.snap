// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UptimeDots renders without errors 1`] = `
<div>
  <div
    class="flex space-x-1"
  >
    <div
      class="h-8 flex-grow bg-green-500 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-green-500 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-red-500 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-green-500 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-red-500 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-red-500 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-green-500 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-red-500 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-green-500 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-red-500 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-red-500 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-green-500 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-green-500 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-gray-200 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-green-500 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-gray-200 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-gray-200 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-gray-200 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-gray-200 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-gray-200 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-gray-200 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-gray-200 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-gray-200 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-gray-200 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-gray-200 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-gray-200 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-gray-200 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-gray-200 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-gray-200 rounded"
      data-testid="uptimeDot"
    />
    <div
      class="h-8 flex-grow bg-gray-200 rounded"
      data-testid="uptimeDot"
    />
  </div>
</div>
`;
